### PR TITLE
[FEATURE] Add hint to adapt addUploadedFile override

### PR DESCRIPTION
**Pull request by helsner**
Resolves: #2916

### CODE REVIEW

**Comment by helsner** [comment]
So this is basically some kind of prototype..
I thought about this change and came to the conclusion that adapting code would be to difficult, as I see the use cases so specific at this point that we cannot provide a generic and working rewrite.
That's why i wanted to at least add a message.

I'm open for your opinions here.
Especially testing this is tricky as there is no real change

File: stubs/TYPO3/CMS/Core/Http/UploadedFile.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+namespace TYPO3\CMS\Core\Http;
+
+if (class_exists('TYPO3\CMS\Core\Http\UploadedFile')) {
+    return;
+}
+
+class UploadedFile
+{
+
+}
```
Comment (simonschaufi) [inline stubs/TYPO3/CMS/Core/Http/UploadedFile.php:11]: new line at EOF missing

File: stubs/TYPO3/CMS/Core/Resource/ResourceStorage.php [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+namespace TYPO3\CMS\Core\Resource;
+
+if (class_exists('TYPO3\CMS\Core\Resource\ResourceFactory')) {
+    return;
+}
+
+class ResourceStorage
+{
+
+}
```
Comment (simonschaufi) [inline stubs/TYPO3/CMS/Core/Resource/ResourceStorage.php:11]: new line at EOF missing

File: tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+-----
+<?php
+
+namespace Ssch\TYPO3Rector\Tests\Rector\v12\v0\typo3\HintNecessaryUploadedFileChangesRector\Fixture;
+
+use TYPO3\CMS\Core\Resource\ResourceStorage;
+
+final class CustomResourceStorage extends ResourceStorage
+{
+    public function addUploadedFile(array $uploadedFileData)
```
Comment (simonschaufi) [inline tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:24]: where is that comment?

File: tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture_no_apply_needed.php.inc [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+<?php
+
+namespace Ssch\TYPO3Rector\Tests\Rector\v12\v0\typo3\HintNecessaryUploadedFileChangesRector\Fixture;
+
+use TYPO3\CMS\Core\Http\UploadedFile;
+use TYPO3\CMS\Core\Resource\ResourceStorage;
+
+final class UpdatedResourceStorage extends ResourceStorage
+{
+    public function addUploadedFile(array|UploadedFile $uploadedFileData)
```
Comment (simonschaufi) [inline tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture_no_apply_needed.php.inc:26]: where is that comment?

**Review by simonschaufi — CHANGES_REQUESTED** [review]

File: tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc (hunk 3 above)
Comment (helsner) [reply tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:24]: I added that comment to enforce the code samples to be valid.
But yes, we could think of the message to be added into the function to hint not only via CLI, but also within the method for the custom changes

File: stubs/TYPO3/CMS/Core/Http/UploadedFile.php (hunk 1 above)
Comment (helsner) [reply stubs/TYPO3/CMS/Core/Http/UploadedFile.php:11]: i had cases where it just wasn't displayed in github? Maybe that's the issue?

File: tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc (hunk 3 above)
Comment (simonschaufi) [reply tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:24]: That would be great. I thought this is the case on other rectors as well but they are all the same. A `TODO: ...` comment would be great.

File: stubs/TYPO3/CMS/Core/Http/UploadedFile.php (hunk 1 above)
Comment (simonschaufi) [reply stubs/TYPO3/CMS/Core/Http/UploadedFile.php:11]: In this case it is really missing.

File: tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc (hunk 3 above)
Comment (helsner) [reply tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:24]: will try that then.

Comment (simonschaufi) [reply tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:24]: I would put the comment at the end of the line if possible and then check if a `TODO: ` comment exists already before adding it.

File: stubs/TYPO3/CMS/Core/Http/UploadedFile.php (hunk 1 above)
Comment (helsner) [reply stubs/TYPO3/CMS/Core/Http/UploadedFile.php]: Yes i got it enabled

File: tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc (hunk 3 above)
Comment (simonschaufi) [reply tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:27]: Easy, bro!

Comment (simonschaufi) [reply tests/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector/Fixture/fixture.php.inc:25]: Ups, I just noticed that I kinda messed up your last force push. Sorry. I adjusted my code according to yours, just slightly different.

**Comment by simonschaufi** [comment]
The failing test is a false positive. It looks like a bug in rector core, but this `EndBug/add-and-commit` workflow is driving me crazy as it is not working at all. @sabbelasichon

File: src/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector.php [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        $params = $affectedMethod->getParams();
+        if ($params[0]->type instanceof UnionType) {
+            return null;
+        }
+
+        if ($params[0]->type instanceof Identifier && 'array' === $params[0]->type->name) {
+            $comments = $affectedMethod->getComments();
+            $comments = array_filter(
+                $comments,
+                static fn (Comment $comment) => ! str_starts_with($comment->getText(), '// FIXME: Rector:')
```
Comment (helsner) [inline src/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector.php:71]: with this check i would expect a test fixture which actively checks this.
What do you think?

Comment (simonschaufi) [reply src/Rector/v12/v0/typo3/HintNecessaryUploadedFileChangesRector.php:71]: done

**Comment by helsner** [comment]
I think it's funny that this should "only be hinting" and now it's bigger than other rules :D 
As PR author i cannot approve, but from my point of view this is fine and can be merged.
Thanks a lot Simon for the rescue!

**Comment by simonschaufi** [comment]
@helsner you can suggest a better name if you come up with something.

**Review by simonschaufi — APPROVED** [review]

**Pull request merged**